### PR TITLE
Add tests for MSC3391

### PR DESCRIPTION
**Pull request by anoadragon453**
This PR adds Complement tests for MSC3391. The MSC is [currently being implemented in Synapse](https://github.com/matrix-org/synapse/pull/14714).

### CODE REVIEW

File: tests/msc3391_test.go [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+	roomID := alice.CreateRoom(t, map[string]interface{}{})
+
+	for _, httpMethod := range []string{"PUT", "DELETE"} {
+		t.Run(fmt.Sprintf("Deleting a user's account data via %s works", httpMethod), func(t *testing.T) {
+			createAndDeleteAccountData(t, alice, httpMethod == "DELETE", nil)
+		})
+		t.Run("Deleting a user's room account data via ", func(t *testing.T) {
+			createAndDeleteAccountData(t, alice, httpMethod == "DELETE", &roomID)
+		})
+	}
```
Comment (clokep) [inline tests/msc3391_test.go:44]: I don't think this loop is really buying us anything TBH.

```suggestion
	// Test deleting global account data.
	t.Run("Deleting a user's account data via DELETE works", func(t *testing.T) {
		createAndDeleteAccountData(t, alice, true, nil)
	})
	t.Run("Deleting a user's account data via PUT works", func(t *testing.T) {
		createAndDeleteAccountData(t, alice, false, nil)
	})

	// Test deleting room account data.
	t.Run("Deleting a user's room data via DELETE works", func(t *testing.T) {
		createAndDeleteAccountData(t, alice, true, &roomID)
	})
	t.Run("Deleting a user's room account data via PUT works", func(t *testing.T) {
		createAndDeleteAccountData(t, alice, false, &roomID)
	})
```

File: tests/msc3391_test.go [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+	// Create the account data and check that it has been created successfully
+	createAccountData(t, c, roomID)
+
+	// Delete the account data and check that it was deleted successfully
+	deleteAccountData(t, c, viaDelete, roomID)
+}
+
+// createAccountData creates some account data for a user or a room, and checks that it was
+// created successfully by both querying the data afterwards, and ensuring it appears down /sync.
+func createAccountData(t *testing.T, c *client.CSAPI, roomID *string) {
```
Comment (clokep) [inline tests/msc3391_test.go:57]: Could this be separate into two functions? It looks like the only duplicated code would be the initial sync?

If you really want to keep a parameter to `createAndDeleteAccountData` for the room ID, I think the if statement could go in there for clarity.

File: tests/msc3391_test.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+// deleteAccountData removes account data for a user or room.
+//
+// If viaDelete is true, a request is made to the DELETE endpoint for user or
+// room account data. Otherwise, the PUT method is used with an empty content
+// dictionary instead. MSC3391 specifies that a PUT with an empty content body
+// is functionally equivalent to deleting an account data type directly.
+//
+// If roomID is not nil, room account data for the given room ID will be removed.
+// Otherwise, account data from the user will be removed instead.
+func deleteAccountData(t *testing.T, c *client.CSAPI, viaDelete bool, roomID *string) {
```
Comment (clokep) [inline tests/msc3391_test.go:150]: Similar question here of if we can separate the global / room functionality.

Comment (anoadragon453) [reply tests/msc3391_test.go]: Good point. I've separated out the various test methods into user and room variants in 58913da.

File: tests/msc3391_test.go [hunk 4]
```diff
@@ -0,0 +1,306 @@
...
+	deleteUserAccountData(t, c, viaDelete)
+}
+
+func createAndDeleteRoomAccountData(t *testing.T, c *client.CSAPI, viaDelete bool, roomID string) {
+	// Create the account data and check that it has been created successfully
+	createRoomAccountData(t, c, roomID)
+
+	// Delete the account data and check that it was deleted successfully
+	deleteRoomAccountData(t, c, viaDelete, roomID)
+}
```
Comment (clokep) [inline tests/msc3391_test.go:68]: I'm not sure these two helpers are that useful, but I won't block on this. 👍

**Review by clokep — APPROVED** [review]

File: tests/msc3391_test.go (hunk 4 above)
Comment (anoadragon453) [reply tests/msc3391_test.go]: Yeah, they really aren't :)

**Pull request merged**